### ISSUE TITLE
Following the exact fine-tune command from README does not give results comparable to the adapter_weights you shared in huggingface

### ISSUE THREAD

**Issue by mriganktiwari** [tloen/alpaca-lora#526]
I too have fine-tuned a Llama-7b with exact same parameters and data as given [here](https://github.com/tloen/alpaca-lora#training-finetunepy), and got a adapter_weights saved.

- When I use these adapter weights saved through  my training, I get something similar as you say, **it keeps repeating the ##Input section of the prompt, without doing anything said in the instruction**.
- This output is very different from what I get when using the adapter weights from HF [repo](https://huggingface.co/tloen/alpaca-lora-7b)

Am unable to understand how can this differ?

_Originally posted by @mriganktiwari in https://github.com/tloen/alpaca-lora/issues/524#issuecomment-1601938432_

**Comment by liyongqi2002** [comment]
> 

I have encountered the same problem, is there a solution?

**Comment by mriganktiwari** [comment]
I tried with lit-llama (from lightning ai) to do the lora fine tuning which worked out fine.

**Comment by Nsigma-Bill** [comment]
Have you checked the size of your adapter_model.bin file? If it is only 443 Bytes, try to remove the following codes when doing finetuning:
```python    old_state_dict = model.state_dict
    model.state_dict = (
        lambda self, *_, **__: get_peft_model_state_dict(
            self, old_state_dict()
        )
    ).__get__(model, type(model))
```

**Comment by meareabc** [comment]
> 

which pre_train model were you use? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.

**Comment by meareabc** [comment]
> > 
> 
> I have encountered the same problem, is there a solution?

which pre_train model were you use? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.

**Comment by meareabc** [comment]
> > 
> 
> I have encountered the same problem, is there a solution?

@liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.

**Comment by Nsigma-Bill** [comment]
> > > 
> > 
> > 
> > I have encountered the same problem, is there a solution?
> 
> @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.

decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.

**Comment by meareabc** [comment]
> > > > 
> > > 
> > > 
> > > I have encountered the same problem, is there a solution?
> > 
> > 
> > @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.
> 
> decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.

Thanks. It seemed that bos_token_id and eos_token_id in decapoda were 0 and 1 respectively. I wonder whether it would perform better if I change  the settings in generate.py to the same as in finetune.py. Because I saw that the loss during finetuning was low, but the model still repeated instructions when generating answers.

**Comment by meareabc** [comment]
> > > > 
> > > 
> > > 
> > > I have encountered the same problem, is there a solution?
> > 
> > 
> > @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.
> 
> decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.

During the previous finetuning, I set bos_token_id to 1 and eos_token_id to 2 in finetune.py, but it did not work well. So I thought if I could just change the settings in generate.py?
I would like to comment out these lines in generate.py as below:
generate.py:
    # model.config.pad_token_id = tokenizer.pad_token_id = 0  # unk
    # model.config.bos_token_id = 1
    # model.config.eos_token_id = 2

**Comment by Nsigma-Bill** [comment]
> > > > > 
> > > > 
> > > > 
> > > > I have encountered the same problem, is there a solution?
> > > 
> > > 
> > > @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.
> > 
> > 
> > decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.
> 
> During the previous finetuning, I set bos_token_id to 1 and eos_token_id to 2 in finetune.py, but it did not work well. So I thought if I could just change the settings in generate.py? I would like to comment out these lines in generate.py as below: generate.py: # model.config.pad_token_id = tokenizer.pad_token_id = 0 # unk # model.config.bos_token_id = 1 # model.config.eos_token_id = 2

Have you checked your adapter_model.bin file size? It should be MB scale rather than Bytes scale. The finetune.py has model saving issues. See my above comment "remove the code".

**Comment by Nsigma-Bill** [comment]
> > > > > 
> > > > 
> > > > 
> > > > I have encountered the same problem, is there a solution?
> > > 
> > > 
> > > @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.
> > 
> > 
> > decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.
> 
> During the previous finetuning, I set bos_token_id to 1 and eos_token_id to 2 in finetune.py, but it did not work well. So I thought if I could just change the settings in generate.py? I would like to comment out these lines in generate.py as below: generate.py: # model.config.pad_token_id = tokenizer.pad_token_id = 0 # unk # model.config.bos_token_id = 1 # model.config.eos_token_id = 2

It is not correct to set `eos_token_id = 0` in order to align with the finetune.py setup. The reason is that in such a case, `eos_token == unknown_token`, the model will not know how to stop generation.

**Comment by liyongqi2002** [comment]
> 

I have solved the problem, I guess the main problem is lora training mode error, I used the following command to fine-tune, the resulting adapter_model.bin is measured in MB, so the output is fine.

`
CUDA_VISIBLE_DEVICES=3 python finetune.py \
    --base_model='decapoda-research/llama-7b-hf' \
    --data_path 'alpaca_data_gpt4.json' \
    --num_epochs=5 \
    --cutoff_len=512 \
    --group_by_length \
    --output_dir='./lora-alpaca' \
    --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' \
    --lora_r=16 \
    --micro_batch_size=8
`

Besides, I also tried to roll back the peft version and modify the eos_token_id in finetune.py, but I think the main reason is probably the lora fine-tuning mode error that leads to insufficient training.

**Comment by Nsigma-Bill** [comment]
> > 
> 
> I have solved the problem, I guess the main problem is lora training mode error, I used the following command to fine-tune, the resulting adapter_model.bin is measured in MB, so the output is fine.
> 
> `CUDA_VISIBLE_DEVICES=3 python finetune.py \ --base_model='decapoda-research/llama-7b-hf' \ --data_path 'alpaca_data_gpt4.json' \ --num_epochs=5 \ --cutoff_len=512 \ --group_by_length \ --output_dir='./lora-alpaca' \ --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' \ --lora_r=16 \ --micro_batch_size=8`
> 
> Besides, I also tried to roll back the peft version and modify the eos_token_id in finetune.py, but I think the main reason is probably the lora fine-tuning mode error that leads to insufficient training.

Hi, could you clarify what you mean by `training mode/lora fine-tuning mode`?

**Comment by meareabc** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > I have encountered the same problem, is there a solution?
> > > > 
> > > > 
> > > > @liyongqi2002 which pre_train model you used? decapoda or yahma? i saw someone said yahma-7b performed better, i would like to use yahma-7b for last try, and then will use lit-llama as above said.
> > > 
> > > 
> > > decapoda has a wrong tokenizer setting (the `bos_token_id`, `eos_token_id` are not correctly assigned to 1 and 2 respectively). On the contrary, yahma has correct settings.
> > 
> > 
> > During the previous finetuning, I set bos_token_id to 1 and eos_token_id to 2 in finetune.py, but it did not work well. So I thought if I could just change the settings in generate.py? I would like to comment out these lines in generate.py as below: generate.py: # model.config.pad_token_id = tokenizer.pad_token_id = 0 # unk # model.config.bos_token_id = 1 # model.config.eos_token_id = 2
> 
> Have you checked your adapter_model.bin file size? It should be MB scale rather than Bytes scale. The finetune.py has model saving issues. See my above comment "remove the code".

I follow your guidence and now my adapter_model.bin is 64MB, but it still didnt work well when i just training 3 epoch in 10k data, which data is not similar to usually talk but some professional tasks.

**Comment by meareabc** [comment]
> > > 
> > 
> > 
> > I have solved the problem, I guess the main problem is lora training mode error, I used the following command to fine-tune, the resulting adapter_model.bin is measured in MB, so the output is fine.
> > `CUDA_VISIBLE_DEVICES=3 python finetune.py \ --base_model='decapoda-research/llama-7b-hf' \ --data_path 'alpaca_data_gpt4.json' \ --num_epochs=5 \ --cutoff_len=512 \ --group_by_length \ --output_dir='./lora-alpaca' \ --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' \ --lora_r=16 \ --micro_batch_size=8`
> > Besides, I also tried to roll back the peft version and modify the eos_token_id in finetune.py, but I think the main reason is probably the lora fine-tuning mode error that leads to insufficient training.
> 
> Hi, could you clarify what you mean by `training mode/lora fine-tuning mode`?

However, when I asked "tell me about alpaca", I observed a difference between the response given by the pretrained model and the LORA. Hence, I presumed it will work better for my professional tasks if i train more epochs.

**Comment by NewEricWang** [comment]
> 
I have used the following command to finetune:
python finetune.py --base_model '../LLM_pretrained_model/llama-7b-hf' --data_path '../LLM_data/trans_chinese_alpaca_data.json' --output_dir './alpaca-lora-zh-v3' --cutoff_len 512 --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' --lora_r=16 --micro_batch_size=8 --num_epochs=5 --group_by_length

The resulting adapter_model.bin is still 433 types.  It almost doesn’t work.

**Comment by liyongqi2002** [comment]
> python finetune.py --base_model '../LLM_pretrained_model/llama-7b-hf' --data_path '../LLM_data/trans_chinese_alpaca_data.json' --output_dir './alpaca-lora-zh-v3' --cutoff_len 512 --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' --lora_r=16 --micro_batch_size=8 --num_epochs=5 --group_by_length


Besides that, you can try this: [https://github.com/tloen/alpaca-lora/issues/334#issuecomment-1509599096](https://github.com/tloen/alpaca-lora/issues/334#issuecomment-1509599096)

**Comment by Nsigma-Bill** [comment]
> > 
> 
> I have used the following command to finetune: python finetune.py --base_model '../LLM_pretrained_model/llama-7b-hf' --data_path '../LLM_data/trans_chinese_alpaca_data.json' --output_dir './alpaca-lora-zh-v3' --cutoff_len 512 --lora_target_modules='[q_proj,k_proj,v_proj,o_proj]' --lora_r=16 --micro_batch_size=8 --num_epochs=5 --group_by_length
> 
> The resulting adapter_model.bin is still 433 types. It almost doesn’t work.
@NewEricWang 
See https://github.com/tloen/alpaca-lora/issues/526#issuecomment-1613636366

**Comment by NewEricWang** [comment]
Thanks. @Nsigma-Bill @liyongqi2002
